numpy/_core/src/umath/loops_hyperbolic.dispatch.cpp.src

```diff
@@ -75,21 +80,103 @@
  *         achieve wider than target precision.
  *
  */
+
+const hn::ScalableTag<float> f32;
+const hn::ScalableTag<int32_t> s32;
+const hn::ScalableTag<uint32_t> u32;
+using vec_f32 = hn::Vec<decltype(f32)>;
+using vec_s32 = hn::Vec<decltype(s32)>;
+using vec_u32 = hn::Vec<decltype(u32)>;
+
+const hn::ScalableTag<double> f64;
+const hn::ScalableTag<int64_t> s64;
+const hn::ScalableTag<uint64_t> u64;
+using vec_f64 = hn::Vec<decltype(f64)>;
+using vec_s64 = hn::Vec<decltype(s64)>;
+using vec_u64 = hn::Vec<decltype(u64)>;
+
+template <typename vtype, typename type_t>
+HWY_ATTR NPY_FINLINE vtype
+load_vector(type_t* src, npy_intp ssrc, npy_intp len){
+    auto D = hn::DFromV<vtype>();
+    using DI = hn::RebindToSigned<decltype(D)>;
+    DI di;
+    
+    auto indices = hn::Iota(di, 0);
+    auto stride = hn::Set(di, ssrc);
+    indices = hn::Mul(indices, stride);
+    
+    const int nlanes = hn::Lanes(D);
+    if (len < nlanes){
+        if (ssrc == 1) {
+            return hn::LoadN(D, src, len);
+        } else {
+            return hn::GatherIndexN(D, src, indices, len); 
+        }
+    }else{
+        if (ssrc == 1) {
+            return hn::LoadU(D, src);
+        } else {
+            return hn::GatherIndex(D, src, indices); 
+        }
+    }
+}
+
+template <typename vtype, typename type_t>
+HWY_ATTR NPY_FINLINE void
+store_vector(vtype vec, type_t* dst, npy_intp sdst, npy_intp len){
+    auto D = hn::DFromV<vtype>();
+    using DI = hn::RebindToSigned<decltype(D)>;
+    DI di;
+    
+    auto indices = hn::Iota(di, 0);
+    auto stride = hn::Set(di, sdst);
+    indices = hn::Mul(indices, stride);
+    
+    const int nlanes = hn::Lanes(D);
+    if (len < nlanes){
+        if (sdst == 1) {
+            hn::StoreN(vec, D, dst, len);
+        } else {
+            hn::ScatterIndexN(vec, D, dst, indices, len); 
+        }
+    }else{
+        if (sdst == 1) {
+            hn::StoreU(vec, D, dst);
+        } else {
+            hn::ScatterIndex(vec, D, dst, indices); 
+        }
+    }
+}
+
 #if NPY_SIMD_F64
 
-    // For architectures without efficient gather / scatter instructions, it is
-    // better to use a transposed LUT where we can load all coefficients for an
-    // index linearly.  In order to keep the same vertical calculation, we
-    // transpose the coef. into lanes.  2 lane transpose is all that's
-    // implemented so we require `npyv_nlanes_f64` == 2.
-    #if npyv_nlanes_f64 == 2
-    #define TANH_TRANSPOSED_LUT
-    #endif // npyv_nlanes_f64 == 2
+HWY_ATTR NPY_FINLINE vec_f64 lut_16_f64(const double * lut, vec_u64 idx){
+    if constexpr(hn::Lanes(f64) == 8){
```

Comment:
This if constexpr will only work `#if !HWY_HAVE_SCALABLE`. We can wrap them, up to the final `} else`, in an #if. Then for SVE, the compiler will only see the { GatherIndex } codepath, which is fine.

Reply:
The reason is: Lanes is only constexpr on some targets, not SVE nor RVV.

Reply:
Or you can use MaxLanes as we do below.

Reply:
I think this is fine for now, as we don't use scalable vectors for this routine (yet).